### DISCUSSION TITLE
Xbox crash

### DISCUSSION THREAD

**Discussion by tarrazuapps** [ffmpeginteropx/FFmpegInteropX#336] [General]
Hello team.

First, I want to thank you for this library, it is really great and you can tell that there is a lot of passion into it.

I wanted to ask if this has this been tested on Xbox? I have been trying the library and works just fine on windows, but on xbox certain streams cause the xbox to crash, it reboots. 

I know I have a long way to go in understanding the internals of the ffmpeg and this library, but do you have any suggestions on how to troubleshoot this?

**Comment by brabebhin** [comment]
Hi,

It hasn't been tested on xbox, as far as I know. I have noticed crashes in my own app on xbox, and have since disabled xbox deployment lol.

I think it has to do with xbox not really fully implementing the UWP contracts. Unfortunately, i don't have an xbox to debug stuff. I'd probably start with attaching a debugger and seeing what's happening.

**Reply by tarrazuapps** [reply]
Thanks for replying.

Since there is no other app in play, why do you think the contract implementation may be at play? I don't fully understand that.

**Reply by lukasf** [reply]
I think there was at least one more dev who reported that xbox crashes or fails on certain codecs. He usedsoftware rendering to work around the problem, but of course it comes at a performance penalty (try MediaSourceConfig.VideoDecoderMode).

My guess is that the GPU drivers do not properly support the HW decoding of some codecs. It is really bad that it just plain crashes.

I also don't have an xbox for testing, so I cannot really help.
Most porbably this is something that only Microsoft could fix.

**Reply by brabebhin** [reply]
There are some pages which show that some features, mostly hardware related, are not available in Xbox. My guess is that they don't fully implement the baseline media codecs required by the UWP contract to save physical space for more shader cores in the GPU. Desktop GPUs usually have dedicated hardware for video encoding and decoding. That means neither pass through nor directx pipelines will work on xbox. Their driver stack is also streamlined and don't fully implement the HAL desktop windows uses. Theoretically the general purpose shader cores can do video encoding and decoding on their own, but probably Microsoft doesn't want to complicate the drivers. 

Honestly, the fact that the whole machine reboots sounds like a serious bug, i dare say it borders the vulnerability realm.

**Reply by tarrazuapps** [reply]
> I think there was at least one more dev who reported that xbox crashes or fails on certain codecs. He usedsoftware rendering to work around the problem, but of course it comes at a performance penalty (try MediaSourceConfig.VideoDecoderMode).
> 
> My guess is that the GPU drivers do not properly support the HW decoding of some codecs. It is really bad that it just plain crashes.
> 
> I also don't have an xbox for testing, so I cannot really help. Most porbably this is something that only Microsoft could fix.

Seems like Config.VideoDecoderMode = VideoDecoderMode.ForceFFmpegSoftwareDecoder; does the trick.
In VLC I also noticed that that videos play better with software instead of hardware acceleration, however, using GPU does not cause the app to crash. 
Is there a way to turn off the GPU acceleration only for certain codecs?

I will do more testing but thank you so much.

**Reply by brabebhin** [reply]
What if you use VideoDecoderMode.SystemDecoder?
This should make use of software decoders when a system codec in not available. The system codec, I assume, should be stable.
This will avoid the D3D11 hardware decoder, which might be the problem.

**Reply by tarrazuapps** [reply]
> What if you use VideoDecoderMode.SystemDecoder? This should make use of software decoders when a system codec in not available. The system codec, I assume, should be stable. This will avoid the D3D11 hardware decoder, which might be the problem.

I tried it and seems to work. This surely sounds like the way to go. Wouldn't it make sense to make it the default, since it is safer?

**Reply by brabebhin** [reply]
The problem with that is that it requires you to download some extensions for certain formats, like hevc, from the windows store. The directx decoder, which is the default, will bypass that and use GPU supplied codecs.

**Reply by lukasf** [reply]
On PC, the behavior is more or less opposite: If you use system passthrough with certain codecs, and the user does not have these codecs installed, video will play with a black image. There is no way to detect if decoding actually worked or not. Plus, Microsoft tries to make money off these codecs, which are in reality just a tiny wrapper around the free to use D3D11 features. When we use ffmpeg D3D11 decoding, we know if it works or not, and we can fallback to software decoding otherwise. And our users do not need to download and install codecs, which is the main reason to go for FFmpeg in the first place.

Since PC is the primary target, D3D11 will be the default decoder pipeline.

**Reply by brabebhin** [reply]
Theoretically we could detect if the device is running xbox, but we would need to take a dependency on microsoft community toolkit to do that. 

https://learn.microsoft.com/en-us/windows/communitytoolkit/helpers/systeminformation

We could also throw an exception when someone uses D3D11 on xbox or change the default to being the passthrough on xbox.

However, I think this responsability should lie with the calling application, not with the library itself.

**Reply by yikuo123** [reply]
> On PC, the behavior is more or less opposite: If you use system passthrough with certain codecs, and the user does not have these codecs installed, video will play with a black image. There is no way to detect if decoding actually worked or not. Plus, Microsoft tries to make money off these codecs, which are in reality just a tiny wrapper around the free to use D3D11 features. When we use ffmpeg D3D11 decoding, we know if it works or not, and we can fallback to software decoding otherwise. And our users do not need to download and install codecs, which is the main reason to go for FFmpeg in the first place.
> 
> Since PC is the primary target, D3D11 will be the default decoder pipeline.

I noticed FFmpeg D3D11 does not work properly on Xbox too, using Automatic gives a green screen.

What is the difference between AutomaticSystemDecoder and ForceSystemDecoder? if we can't detect if system decoding is successful?

I noticed that some hevc videos can be decoded with Automatic or ForceFFmpegSoftwareDecoder mode on PC, but both AutomaticSystemDecoder and ForceSystemDecoder will have a black screen. However, it works with MediaPlayer without using FFmpegMediaSource.
